File: Week1/hard/接雨水/Solution.java
package 接雨水;
/*
执行用时：1 ms, 在所有 Java 提交中击败了99.98%的用户
内存消耗：38.2 MB, 在所有 Java 提交中击败了76.59%的用户
* */
class Solution {
    public static void main(String[] args) {
        int[] h = {0,1,0,2,1,0,1,3,2,1,2,1};
        System.out.println(trap(h));
    }
    public static int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int leftMax = 0;
        int rightMax = 0;
        int sum = 0;
        while (left < right){
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (leftMax <= rightMax){
                sum += leftMax - height[left++];
            }else {
                sum += rightMax - height[right--];
            }
        }
        return sum;
    }
}

File: Week1/hard/接雨水/Solution.js
/*
执行用时：84 ms, 在所有 JavaScript 提交中击败了94.07%的用户
内存消耗：39.4 MB, 在所有 JavaScript 提交中击败了42.68%的用户
* */

/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = 0, right = height.length - 1,
        leftMax = 0, rightMax = 0 ,sum = 0;
    while (left < right){
        leftMax = Math.max(leftMax, height[left]);
        rightMax = Math.max(rightMax, height[right]);
        if (leftMax <= rightMax)sum += leftMax - height[left++];
        else sum += rightMax - height[right--];
    }
    return sum;
};